### PR TITLE
Refactor filterParameters state to use search params state shape and filter config

### PR DESCRIPTION
**Pull request by mddilley**
## Associated issues
https://github.com/cityofaustin/atd-data-tech/issues/15049
https://github.com/cityofaustin/atd-data-tech/issues/15074

This PR:
- updates the `filterParameters` state to have the same shape and key-value pairs (field, operator, and value) as the `filters` state that powers the query
- moves a lot of functions defined within the `Filters` component into a helpers file
- removes code that used to access configuration about a field and then store it in state so that it now accessing that configuration as needed

If you want to visualize the biggest change, run locally and log `filterParameters` in the `Filters` component as you add advanced filters. Compare what that looks like with this branch versus `main` branch.

## Testing
**URL to test:** 

https://deploy-preview-1236--atd-moped-main.netlify.app/moped/projects

**Steps to test:**
1. Go to the project list view and add advanced filters
2. Test a filter with lookup tables like **Type**
3. Test editing applied filters options (choose new field option, choose new operator option, or update search value)
4. Test deleting filters with the trash can icon. A new search should not happen until you click the Search button.
5. Test clearing the first **Field** input after selecting a value by using the X icon that appears to the left of the dropdown icon. Both the **Field** and **Operator** values should clear from the UI.
6. Test resetting the filters by clicking the reset button
7. Test making edits to the url string to make sure the app doesn't break when it can't parse the URL params
8. Test filtering, clicking on a project name link in the list view, and then clicking the "< ALL PROJECTS" link to go back to the list view. Your filters should still be present.
9. Add a filter that expects a number value like **ID is** and try entering letters. You should see that the recently-added validation still works.
10. Check that the trash can delete icons and match any/all radios do not appear until more than one filter is added.

---
#### Ship list
- [ ] Code reviewed 
- [ ] Product manager approved
- [ ] Added to [QA test script](https://docs.google.com/spreadsheets/d/1n_O6MLh9cwwPf57HUM394Ea-z9uuoEV1-QW4axNZXLE/edit#) if applicable

### CODE REVIEW

File: moped-editor/src/components/GridTable/Filters.js [hunk 1]
```diff
@@ -1,6 +1,5 @@
-import React, { useCallback, useEffect, useState, useMemo } from "react";
+import React, { useCallback, useState, useMemo } from "react";
 import { useSearchParams } from "react-router-dom";
-import { v4 as uuidv4 } from "uuid";
```
Comment (mddilley) [inline moped-editor/src/components/GridTable/Filters.js:3]: Since the shape of `filterParameters` is now an array, we can use the index of a filter in the array to point to the filter to edit, delete, etc. Previously, we used an object so we needed to assign a unique ID so that we could point to the right filter in the object.

File: moped-editor/src/components/GridTable/Filters.js [hunk 2]
```diff
@@ -96,107 +96,6 @@ const useStyles = makeStyles((theme) => ({
...
-  return { ...defaultNewFieldState, id: uuid };
-};
-
-/**
- * Get configuration about the filters using the query parameters present in the URL
- * @param {Array} filters - filters from the URL {field, operator, value}
- * @param filtersConfig - The configuration object for the filters
- * @return {Object}
- */
-const makeInitialFilterParameters = (filters, filtersConfig) => {
```
Comment (mddilley) [inline moped-editor/src/components/GridTable/Filters.js:141]: We don't need this helper anymore. It took the three values stored in the search parameters (field, operator, and value) and built up the shape of data that the UI expected from our predefined `FiltersCommonOperators` and `PROJECT_LIST_VIEW_FILTERS_CONFIG`. It was a bridge to the UI changes made in this PR. Most of this work happens in the render now.

File: moped-editor/src/components/GridTable/Filters.js [hunk 3]
```diff
@@ -230,215 +129,108 @@ const Filters = ({
...
-      AUTOCOMPLETE_OPERATORS.includes(field.operator)
-    );
-  };
+  /* Some features like all/any radios require more than one filter to appear */
+  const areMoreThanOneFilters = filterParameters.length > 1;
 
   /**
    * Handles the click event on the field drop-down menu
-   * @param {string} filterId - State FieldID to modify
+   * @param {string} filterIndex - filterParameters index to modify
```
Comment (mddilley) [inline moped-editor/src/components/GridTable/Filters.js:160]: I renamed `filterId` throughout this component to `filterIndex`. Hoping that makes it clearer that we are dealing with array indices instead of the uuids.

File: moped-editor/src/components/GridTable/Filters.js [hunk 4]
```diff
@@ -230,215 +129,109 @@ const Filters = ({
...
-
-      // Find the field we need to gather options from
-      const fieldDetails = filtersConfig.fields.find(
-        (filter) => filter.name === field
-      );
-
-      if (!fieldDetails) {
-        filtersNewState[filterId] = generateEmptyField(filterId);
-      } else {
-        // Update field & type
```
Comment (mddilley) [inline moped-editor/src/components/GridTable/Filters.js:310]: From here and below in this `handleFilterFieldMenuClick` function, the code was building up new state (similar to what `makeInitialFilterParameters` above was doing). This happens in the return below where `filterParameters.map` happens. The available operators logic here is captured in a new `getAvailableOperators` helper.

File: moped-editor/src/components/GridTable/Filters.js [hunk 5]
```diff
@@ -230,215 +129,109 @@ const Filters = ({
...
+    const fieldDetails = filtersConfig.fields.find(
+      (filter) => filter.name === field
+    );
+
+    // Update the field and operator values or handle when the clear X icon is clicked
+    if (fieldDetails) {
+      filtersNewState[filterIndex].field = fieldDetails.name;
+      filtersNewState[filterIndex].operator = getDefaultOperator(fieldDetails);
     } else {
-      console.debug(
```
Comment (mddilley) [inline moped-editor/src/components/GridTable/Filters.js:368]: imo this debug message isn't needed any longer since the number of filter rows in the UI and the number of items in the `filterParameters` state is equal. There should always be a corresponding filter to a filter row being interacted with.

File: moped-editor/src/components/GridTable/Filters.js [hunk 6]
```diff
@@ -481,103 +273,26 @@ const Filters = ({
...
-   */
-  const isFilterNullType = (field) => {
-    return field.gqlOperator && field.gqlOperator.includes("is_null");
-  };
-
-  /**
-   * This side effect monitors the count of filters
-   * if the count of filters is zero, then resets the state
-   */
-  useEffect(() => {
```
Comment (mddilley) [inline moped-editor/src/components/GridTable/Filters.js:554]: Making sure that an empty filter is in `filterParameters` state now happens on first render when `initialFilterParameters` is determined. Also, since the trash can icon disappears when there is only one filter, we don't have to handle the filter count dropping to 0 that way.

File: moped-editor/src/components/GridTable/Filters.js [hunk 7]
```diff
@@ -481,103 +273,26 @@ const Filters = ({
...
-      handleClearFilters();
-      // Add an empty filter so the user doesn't have to click the 'add filter' button
-      generateEmptyFilter();
-    }
-  }, [filterParameters, setFilters, generateEmptyFilter, handleClearFilters]);
-
-  /**
-   * This side effect monitors whether the user has added a complete filter
-   */
-  useEffect(() => {
```
Comment (mddilley) [inline moped-editor/src/components/GridTable/Filters.js:566]: The `filterComplete` state is removed and replaced with the `areAllFiltersComplete` helper. We can use that function to check if all filters are complete during render since it is calculated from `filterParameters`.

**Review by johnclary — CHANGES_REQUESTED** [review]
😍 so much red!

this is such a major improvement, Mike! i ran into one regression, and besides that my comments were pretty minor.

File: moped-editor/src/components/GridTable/Filters.js [hunk 8]
```diff
@@ -637,14 +352,33 @@ const Filters = ({
...
-        const isValidInput = checkIsValidInput(filterParameters[filterId]);
+      {filterParameters.map((filter, filterIndex) => {
+        const { field: fieldName, operator, value } = filter;
+
+        /* Get field configuration and values need for inputs */
+        const fieldConfig = filtersConfig.fields.find(
+          (field) => field.name === fieldName
+        );
+        const { label, type } = fieldConfig ?? {};
+        const operators = fieldConfig?.operators;
```
[the remark this answers is not in the document]
Comment (mddilley) [reply moped-editor/src/components/GridTable/Filters.js:363]: nah not nits to me - this is great stuff and thank you for taking such a detailed look at this! 🙏 I'll address these comments today.

File: moped-editor/src/components/GridTable/Filters.js [hunk 9]
```diff
@@ -230,215 +129,109 @@ const Filters = ({
...
+      // if we are switching to an autocomplete input, clear the search value
+      if (shouldRenderAutocompleteInput(lookupTable, operator, loading)) {
+        filtersNewState[filterIndex].value = null;
+      }
     } else {
-      console.debug(
-        `The filter id ${filterId} does not exist, ignoring click event.`
-      );
+      // Reset operator value
+      filtersNewState[filterIndex].operator = null;
```
[the remark this answers is not in the document]
Comment (frankhereford) [reply moped-editor/src/components/GridTable/Filters.js:205]: Maybe if the user is using the dev tools to modify the request on the way the server or Postman to create malformed ones for devious intent? I don't think it's something we need to worry about too much, but it did come to mind wrt your question.

**Review by frankhereford — APPROVED** [review]
Mike, I have gone over your code, but there is a lot of complexity going on here with a system that I have never really had a chance to work on, so I don't feel really qualified to be able to give you any meaningful feedback, beyond that it looks great. Nothing stuck out to me as even remotely out of place in your work.

I was able to test the deploy preview, and I wasn't able to get it to do anything that it shouldn't. The behavior all felt natural and expected, and nothing broke. I think this is ship shape, and I'm really grateful for this huge amount of work you've done!

File: moped-editor/src/components/GridTable/Filters.js (hunk 9 above)
[the remark this answers is not in the document]
Comment (chiaberry) [reply moped-editor/src/components/GridTable/Filters.js:205]: I'm not sure how that would work, I dont see how the operator would not be in the list of operators, since the select dropdown is populated based on that list of operators. Directly editing the operator in the url to something else doesn't affect this event handler either

File: moped-editor/src/components/GridTable/Filters.js (hunk 7 above)
Comment (mddilley) [reply moped-editor/src/components/GridTable/Filters.js:566]: This is interesting - i think that you might be on to something. I fixed this by adding the two operators that this filters uses ("subprojects_array_is_null" and "subprojects_array_is_not_null") to the list the code looks at when determining whether to show the value input or not (`OPERATORS_WITHOUT_SEARCH_VALUES`). Somehow the prod code handles this case without this change. 🤔

File: moped-editor/src/components/GridTable/Filters.js [hunk 10]
```diff
@@ -230,215 +129,109 @@ const Filters = ({
...
-        if (renderAutocompleteInput(filtersNewState[filterId])) {
-          filtersNewState[filterId].value = null;
-        }
-      } else {
-        // Reset operator values
-        filtersNewState[filterId].operator = null;
-        filtersNewState[filterId].gqlOperator = null;
-        filtersNewState[filterId].envelope = null;
-      }
+  const handleFilterOperatorClick = (filterIndex, operator, lookupTable) => {
```
[the remark this answers is not in the document]
Comment (mddilley) [reply moped-editor/src/components/GridTable/Filters.js:191]: great call - i'll update them. 👍

File: moped-editor/src/components/GridTable/Filters.js (hunk 9 above)
[the remark this answers is not in the document]
Comment (mddilley) [reply moped-editor/src/components/GridTable/Filters.js:205]: This is a really good catch because this code used to pull double duty and call `handleFilterOperatorChange` from within `handleFilterFieldMenuChange` to reset the operator input's value. So, this block was meant to handle that case. 

TLDR It used to be able to happen, but it cannot anymore. I'll remove it. 🗑️  Thanks, y'all!

File: moped-editor/src/components/GridTable/Filters.js (hunk 7 above)
Comment (mddilley) [reply moped-editor/src/components/GridTable/Filters.js:566]: Added "string_is_blank" and "string_is_not_blank" to that list as well to fix broken **Status is blank** or **Status is not blank** 😑

File: moped-editor/src/components/GridTable/Filters.js (hunk 9 above)
[the remark this answers is not in the document]
Comment (mddilley) [reply moped-editor/src/components/GridTable/Filters.js]: This also helped me unearth an issue when switching from existing filter with a search value like **Names contains congress** to one without a search value like **Lead is not blank**. The search value was hanging around in the search params so I made another code update. 🙏 🙏 🙏

File: moped-editor/src/components/GridTable/Filters.js [hunk 11]
```diff
@@ -230,215 +129,109 @@ const Filters = ({
...
-    };
-    // Try to delete the filter by filterId
+  const handleDeleteFilterButtonClick = (filterIndex) => {
+    // Clone the state
+    const filtersNewState = [...filterParameters];
+    // Try to delete the filter by filter index
     try {
       // Delete the key (if it's there)
-      delete filtersNewState[filterId];
+      filtersNewState.splice(filterIndex, 1);
```
[the remark this answers is not in the document]
Comment (mddilley) [reply moped-editor/src/components/GridTable/Filters.js:237]: Yup! I checked and this can go and is now gone 👋

File: moped-editor/src/components/GridTable/Filters.js (hunk 8 above)
[the remark this answers is not in the document]
Comment (mddilley) [reply moped-editor/src/components/GridTable/Filters.js:366]: This took me a second, but, yes, I see it now. Yeah, I overlooked how the list of operators are keys in `FiltersCommonOperators`. I'm going to update so that the form input just accesses the label name using the operator key. I can get rid of a helper this way too. 😈

**Review by johnclary — CHANGES_REQUESTED** [review]
@mddilley this batch of changes looks really good—thank you!

i noticed an annoying regression, which does not seem to occur in production currently. The filters break if you switch the filter field from a text-based to a number-based field. For example, create a filter on **Name** **contains** **decker**, then switch the field to **ID**.

Screenshot 2024-01-11 at 12 10 13 PM

you mind taking a look at that?

**Comment by mddilley** [comment]
@johnclary yup, this is a regression related to using * in the `PROJECT_LIST_VIEW_FILTERS_CONFIG` field config to represent [using "all the filters for a specific type"](https://github.com/cityofaustin/atd-moped/pull/1236/commits/205cb33acf49357a44bbca1e5026f7fc278e16a6#diff-e290922fa8581df0703a6896695b704d451d814d5eb327082744c66e960d5afeL139-L148). I'm going to make the two fields that use this shortcut more explicit. `getAvailableOperators` handled this special case but I removed it. 💥 🙃

**Review by johnclary — APPROVED** [review]
Yep—this is ready to ship! Thanks @mddilley for all the iterations. 

You love to see it:
Screenshot 2024-01-16 at 12 55 27 PM

🚢 🚢 🚢

**Comment by mddilley** [comment]
Going to merge this one and add a note to [this thread](https://github.com/cityofaustin/atd-moped/pull/1241#discussion_r1453791101) since @tillyw was going to need to merge this branch into that PR's branch.
